packages/devtools_app_shared/lib/src/service/service_extension_manager.dart

```diff
@@ -254,68 +254,90 @@ final class ServiceExtensionManager with DisposerMixin {
       }
     } else {
       // Set any extensions that are already enabled on the device. This will
       // enable extension states in DevTools on page refresh or initial start.
-      return await _restoreExtensionFromDevice(name);
+      final restored = await _restoreExtensionFromDeviceIfReady(name);
+      if (restored) {
+        // Only mark `name` as an "added service extension" if it was truly
+        // restored. If it was restored, then subsequent calls to
+        // `_addServiceExtension` with `name` will return early. If it was not
+        // really restored, then subsequent calls to `_addServiceExtension`
+        // with `name` will proceed as usual.
+        _serviceExtensions.add(name);
+      }
     }
   }
 
   IsolateRef? get _mainIsolate => _isolateManager.mainIsolate.value;
 
-  Future<void> _restoreExtensionFromDevice(String name) async {
+  /// Restores the service extension named [name] from the device.
+  ///
+  /// Returns whether isolates in the connected app are prepared for the restore.
+  Future<bool> _restoreExtensionFromDeviceIfReady(String name) async {
     final isolateRef = _isolateManager.mainIsolate.value;
-    if (isolateRef == null) return;
+    if (isolateRef == null) return false;
 
     if (!extensions.serviceExtensionsAllowlist.containsKey(name)) {
-      return;
+      return true;
     }
     final expectedValueType =
         extensions.serviceExtensionsAllowlist[name]!.values.first.runtimeType;
 
-    Future<void> restore() async {
+    /// Restores the service extension named [name].
+    ///
+    /// Returns whether isolates in the connected app are prepared for the
+    /// restore.
+    Future<bool> restore() async {
       // The restore request is obsolete if the isolate has changed.
-      if (isolateRef != _mainIsolate) return;
+      if (isolateRef != _mainIsolate) return false;
       try {
         final response = await _service!.callServiceExtension(
           name,
           isolateId: isolateRef.id,
         );
 
-        if (isolateRef != _mainIsolate) return;
+        if (isolateRef != _mainIsolate) return false;
 
         switch (expectedValueType) {
           case const (bool):
             final enabled = response.json!['enabled'] == 'true' ? true : false;
             await _maybeRestoreExtension(name, enabled);
-            return;
           case const (String):
             final String? value = response.json!['value'];
             await _maybeRestoreExtension(name, value);
-            return;
           case const (int):
           case const (double):
             final value = num.parse(
               response.json![name.substring(name.lastIndexOf('.') + 1)],
             );
             await _maybeRestoreExtension(name, value);
-            return;
           default:
-            return;
+            return true;
         }
       } catch (e) {
         // Do not report an error if the VMService has gone away or the
```

Comment:
should we still return false on catch? or at least when the error indicates the service is disposed? See `isServiceDisposedError` helper

Reply:
Yeah we can do:
```
on RPCError catch (e) {
  if (e.isServiceDisposedError) {
    return false;
  }
} catch (e) {
  // swallow
}
```